Compile/src/Producers/LeptonSFProducer.cc

```diff
@@ -35,15 +38,133 @@ std::string LeptonSFProducer::GetProducerId() const { return "LeptonSFProducer";
+              << histoname;
+    TFile file(m_sffile.c_str(), "READONLY");
+    if(!file.Get(histoname.c_str())) {
+        LOG(FATAL) << "Failed to open histogram " << histoname << " from file " << m_sffile;
+    }
+    sfhisto = (TH2F*)file.Get(histoname.c_str());
+    sfhisto->SetDirectory(0);
+    file.Close();
+
+    // use this pointer, so we call child class methods if overwritten.
+    // set correct bools for further processing
+    this->SetEtaAxis2D(histoname);
+    // Use highest pT bin as overflow bin
+    this->SetOverflowPtBin();
 }
 
 void LeptonSFProducer::Produce(ZJetEvent const& event,
-                                 ZJetProduct& product,
-                                 ZJetSettings const& settings) const
+                               ZJetProduct& product,
+                               ZJetSettings const& settings) const
 {
-    
+    LOG(DEBUG) << "\n[" << this->GetProducerId() << "]";
+    if (product.m_zValid) {
+        const auto* lep1 = product.m_originalLeptons.at(product.m_zLeptons.first);
+        const auto* lep2 = product.m_originalLeptons.at(product.m_zLeptons.second);
+        LOG(DEBUG) << "Apply Variation: " << settings.GetLeptonSFVariation();
+        float sf1 = 1 / this->GetScaleFactor(0, *lep1);
+        float sf2 = 1 / this->GetScaleFactor(0, *lep2);
+        product.m_optionalWeights["zl1" + weightName] = sf1;
+        product.m_optionalWeights["zl2" + weightName] = sf2;
+        product.m_weights["lepton" + weightName] = sf1 * sf2;
+        LOG(DEBUG) << "Lepton1 SF: " << sf1 << ", Lepton2 SF: " << sf2;
+        LOG(DEBUG) << "Total SF: " << sf1 * sf2;
+        if (settings.GetLeptonSFVariation() == true) {
+            // up has -1 since it's inside the denominator, down +1
+            float sf1u = 1 / this->GetScaleFactor(-1, *lep1);
+            float sf1d = 1 / this->GetScaleFactor(+1, *lep1);
+            float sf2u = 1 / this->GetScaleFactor(-1, *lep2);
+            float sf2d = 1 / this->GetScaleFactor(+1, *lep2);
+            product.m_optionalWeights["zl1" + weightName + "Up"] = sf1u;
+            product.m_optionalWeights["zl1" + weightName + "Down"] = sf1d;
+            product.m_optionalWeights["zl2" + weightName + "Up"] = sf2u;
+            product.m_optionalWeights["zl2" + weightName + "Down"] = sf2d;
+            // proper error propagation needed
+            product.m_optionalWeights["lepton" + weightName + "Up"] = sf1u * sf2u;
+            product.m_optionalWeights["lepton" + weightName + "Down"] = sf1d * sf2d;
+            if (settings.GetDebugVerbosity() > 1) {
+                LOG(DEBUG) << "Lepton1 SF up: " << sf1u << ", Lepton1 SF down: " << sf1d;
+                LOG(DEBUG) << "Lepton2 SF up: " << sf2u << ", Lepton2 SF down: " << sf2d;
+                LOG(DEBUG) << "Total SF up: " << sf1u * sf2u << ", Total SF down: " << sf1d * sf2d;
+            }
+        }
+    } else {
+        LOG(DEBUG) << "No valid Z found. Only applying SFs to Z leptons.";
```

Comment:
What is ment with "Z leptons" in this case?
Since no valid Z is available?

Reply:
Yes, I only apply the SFs to the leptons from a Valid Z. Applying them to a third valid Muon is wrong, as we only care about the two Z muons.

Reply:
But if m_validZ is false, we have no "Z leptons". Therefore, I find this Debug message missleading.

Reply:
True, and then No SFs are calculated and stored, which is what I want. The default value for weights, if they are not set is 1. Which should be fine.